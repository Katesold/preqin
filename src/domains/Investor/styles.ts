import styled from "styled-components";

export const InvestorContainer = styled.div`
  position: relative;
  display: inline-block;
  text-align: center;
  h1 {
    padding: 20px;
  }
`;
