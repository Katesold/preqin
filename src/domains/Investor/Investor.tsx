import { useEffect, useState } from "react";

import { AssetClasses } from "../../types";
import Dropdown from "../../components/Dropdown/Dropdown";
import { InvestorContainer } from "./styles";
import { StyledTable } from "../../components/Table/styles";
import Table from "../../components/Table/Table";
import { useLocation } from "react-router-dom";
import { useQuery } from "@tanstack/react-query";

export const Investor = () => {
  const options = Object.values(AssetClasses);
  const location = useLocation();
  const [selectedOption, setSelectedOption] = useState("Select an option");
  const [tableData, setTableData] = useState([]);

  const fetchCommitment = async (investorId = 1, selectedOption: string) => {
    const assetClass = selectedOption.slice(0, 2);
    const asset = Object.keys(AssetClasses).find((a) => a.includes(assetClass));

    const response = await fetch(
      `https://api.preqin.com/api/Investor/commitment/${
        asset ? asset.toLowerCase() : ""
      }/${investorId}`,
      {
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${process.env.REACT_APP_TOKEN}`,
        },
        method: "GET",
      }
    );
    if (!response.ok) {
      throw new Error("Failed to fetch insvestor commitment");
    }

    return response.json();
  };

  const { data, isLoading, isError, error, isFetching } = useQuery({
    queryKey: ["commitment"],
    queryFn: () => fetchCommitment(location.state.firmID, selectedOption),
    enabled: !!location?.state?.firmID && selectedOption !== "Select an option",
  });

  const handleSelect = (option: string) => {
    setSelectedOption(option);
  };

  useEffect(() => {
    if (isError && error) {
      window.alert(error);
    }
  }, [isError, error]);

  useEffect(() => {
    if (selectedOption === "Select an option" || isError) {
      setTableData([]);
    } else {
      setTableData(data?.data);
    }
  }, [selectedOption, data?.data, isError]);

  return (
    <InvestorContainer>
      <h1>Investor {location?.state?.firmName}</h1>
      {location?.state ? <Table firms={[location?.state]} readOnly /> : null}
      <Dropdown
        options={options}
        onSelect={handleSelect}
        selectedOption={selectedOption}
      />
      {isLoading || isFetching ? (
        <p>Loading..</p>
      ) : (
        <div>
          <h2 style={{ padding: "20px" }}>Fund Details</h2>
          <StyledTable style={{ width: "50%", margin: "auto" }}>
            <tbody>
              {tableData?.length
                ? tableData.map(
                    (investor: { [s: string]: unknown } | ArrayLike<unknown>) =>
                      Object.entries(investor).map(([key, value]) => (
                        <tr key={key + value}>
                          <td
                            style={{
                              padding: "10px",
                            }}
                          >
                            {key}
                          </td>
                          <td>{value as string} </td>
                        </tr>
                      ))
                  )
                : null}
            </tbody>
          </StyledTable>
        </div>
      )}
    </InvestorContainer>
  );
};
