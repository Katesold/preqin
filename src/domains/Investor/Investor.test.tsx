import "@testing-library/jest-dom";

import {
  RouterProvider,
  createMemoryRouter,
  useLocation,
} from "react-router-dom";
import { fireEvent, render, screen } from "@testing-library/react";

import { Investor } from "./Investor";
import { useQuery } from "@tanstack/react-query"; // Mock react-query's useQuery

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useLocation: jest.fn(),
}));

jest.mock("@tanstack/react-query", () => ({
  ...jest.requireActual("@tanstack/react-query"),
  useQuery: jest.fn(),
}));

describe("Investor Component", () => {
  const mockLocationState = {
    firmID: "123",
    firmName: "Test Firm",
    firmType: "type",
    yearEst: "",
    address: "address 123",
  };

  beforeEach(() => {
    (useQuery as any).mockReturnValue({
      data: {
        data: [
          {
            investorId: "2670",
            investorName: "1199SEIU National Benefit Fund",
            fundId: "16932",
            fundName: "AFL-CIO Housing Investment Trust",
            fundManagerId: "12699",
            fundManagerName: "AFL-CIO",
            fundCurrency: "",
            fundSizeMn: "",
            committedMn: "",
            vintage: "2010",
            fundType: "Real Estate Debt",
            primarySector: "Residential",
            domicile: "Maryland",
            benchmarkLocations: "North America",
            managerExperience: "",
          },
        ],
      },
      isLoading: false,
      isError: false,
      error: null,
      refetch: jest.fn(),
    });
    (useLocation as any).mockReturnValue({
      state: mockLocationState,
    });
  });

  it("renders investor component with initial state", async () => {
    const routes = [
      {
        path: "investors/:Id",
        element: <Investor />,
      },
    ];

    const router = createMemoryRouter(routes, {
      initialEntries: ["/investors/:Id"],
      initialIndex: 0,
    });
    const { container } = render(<RouterProvider router={router} />);

    expect(screen.getByText(/Firm ID/)).toBeInTheDocument();
    fireEvent.click(screen.getByText("Select an option"));
    fireEvent.click(screen.getByText("PE(Private Equity)"));
    expect(
      screen.getByText(/1199SEIU National Benefit Fund/)
    ).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
