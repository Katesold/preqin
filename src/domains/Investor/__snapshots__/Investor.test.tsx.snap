// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Investor Component renders investor component with initial state 1`] = `
<div>
  <div
    class="sc-jTQCzO hIlhHy"
  >
    <h1>
      Investor 
      Test Firm
    </h1>
    <table
      class="sc-iBdnpw bqszkd"
    >
      <thead
        class="sc-fsYfdN fQScNy"
      >
        <tr
          class="sc-kFCroH ebMFOg"
        >
          <th
            class="sc-jsEeTM clubGg"
          >
            Firm ID
          </th>
          <th
            class="sc-jsEeTM clubGg"
          >
            Firm Name
          </th>
          <th
            class="sc-jsEeTM clubGg"
          >
            Type
          </th>
          <th
            class="sc-jsEeTM clubGg"
          >
            Date Added
          </th>
          <th
            class="sc-jsEeTM clubGg"
          >
            Address
          </th>
        </tr>
      </thead>
      <tbody
        class="sc-qZrbh gUswMD"
        readonly=""
      >
        <tr
          class="sc-irLvIq euuTWe"
        >
          <td
            class="sc-csKJxZ bpPyNX"
          >
            123
          </td>
          <td
            class="sc-csKJxZ bpPyNX"
          >
            Test Firm
          </td>
          <td
            class="sc-csKJxZ bpPyNX"
          >
            type
          </td>
          <td
            class="sc-csKJxZ bpPyNX"
          >
            -
          </td>
          <td
            class="sc-csKJxZ bpPyNX"
          >
            address 123
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="sc-beySPh knomkA"
    >
      <div
        class="sc-guDLey jehYId"
      >
        PE(Private Equity)
        <span
          class="sc-dmyCSP dLJUjY"
        />
      </div>
    </div>
    <div>
      <h2
        style="padding: 20px;"
      >
        Fund Details
      </h2>
      <table
        class="sc-iBdnpw bqszkd"
        style="width: 50%; margin: auto;"
      >
        <tbody>
          <tr>
            <td
              style="padding: 10px;"
            >
              investorId
            </td>
            <td>
              2670
               
            </td>
          </tr>
          <tr>
            <td
              style="padding: 10px;"
            >
              investorName
            </td>
            <td>
              1199SEIU National Benefit Fund
               
            </td>
          </tr>
          <tr>
            <td
              style="padding: 10px;"
            >
              fundId
            </td>
            <td>
              16932
               
            </td>
          </tr>
          <tr>
            <td
              style="padding: 10px;"
            >
              fundName
            </td>
            <td>
              AFL-CIO Housing Investment Trust
               
            </td>
          </tr>
          <tr>
            <td
              style="padding: 10px;"
            >
              fundManagerId
            </td>
            <td>
              12699
               
            </td>
          </tr>
          <tr>
            <td
              style="padding: 10px;"
            >
              fundManagerName
            </td>
            <td>
              AFL-CIO
               
            </td>
          </tr>
          <tr>
            <td
              style="padding: 10px;"
            >
              fundCurrency
            </td>
            <td>
               
            </td>
          </tr>
          <tr>
            <td
              style="padding: 10px;"
            >
              fundSizeMn
            </td>
            <td>
               
            </td>
          </tr>
          <tr>
            <td
              style="padding: 10px;"
            >
              committedMn
            </td>
            <td>
               
            </td>
          </tr>
          <tr>
            <td
              style="padding: 10px;"
            >
              vintage
            </td>
            <td>
              2010
               
            </td>
          </tr>
          <tr>
            <td
              style="padding: 10px;"
            >
              fundType
            </td>
            <td>
              Real Estate Debt
               
            </td>
          </tr>
          <tr>
            <td
              style="padding: 10px;"
            >
              primarySector
            </td>
            <td>
              Residential
               
            </td>
          </tr>
          <tr>
            <td
              style="padding: 10px;"
            >
              domicile
            </td>
            <td>
              Maryland
               
            </td>
          </tr>
          <tr>
            <td
              style="padding: 10px;"
            >
              benchmarkLocations
            </td>
            <td>
              North America
               
            </td>
          </tr>
          <tr>
            <td
              style="padding: 10px;"
            >
              managerExperience
            </td>
            <td>
               
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
